PyRoute/TradeCodes.py

```diff
@@ -80,6 +96,148 @@ def __init__(self, initial_codes):
+                continue
+            if 7 < len(raw) and '(' == raw[0] and ')' == raw[-2]:  # Let older-style sophont codes through
+                codes.append(raw)
+                continue
+            if 2 == len(raw) and ('W' == raw[1] or raw[1].isdigit()):
+                if 'C' == raw[0]:  # Compact chirper population
+                    pop = raw[1]
+                    codes.append('Chir' + pop)
+                    continue
+                if 'D' == raw[0]:  # Compact Droyne population
+                    pop = raw[1]
+                    codes.append('Droy' + pop)
+                    continue
+            if not raw.startswith('(') and '(' in raw and raw.endswith(')'):
+                continue
+            if not raw.endswith(')') and ')' in raw and raw.startswith('(') and 7 > len(raw):
+                continue
+            if not raw.startswith('(') and not raw.endswith(')') and '(' in raw and ')' in raw:
+                continue
+            if 7 == len(raw) and '(' == raw[0] and ')' == raw[5]:  # Let preprocessed sophont codes through
+                codes.append(raw)
+                continue
+            if not raw.startswith('('):
+                codes.append(raw)
+                continue
+            if raw.endswith(')'):
+                codes.append(raw)
+                continue
+            if i < num_codes - 1:  # Try to join cases like '(Carte' and 'Blanche)'
+                next = raw_codes[i + 1]
+                if next.endswith(')'):
+                    combo = raw + ' ' + next
+                    codes.append(combo)
+                    raw_codes[i + 1] = ''
+                continue
+
+        codes = sorted(codes)
+
+        initial_codes = ' '.join(codes)
+
+        return codes, initial_codes
+
+    def _process_homeworld(self, homeworld, homeworlds_found, initial_codes):
+        full_name = re.sub(r'\(([^)]+)\)\d?', r'\1', homeworld)
+        homeworlds_found.append(homeworld)
+        match = TradeCodes.search.match(homeworld)
+        if match is None:
+            self.logger.error("Unable to process %s", initial_codes)
+            sys.exit(1)
+        if full_name.startswith("Di"):
+            code = match.group(1)
+            pop = 'X'
+        else:
+            code = match.group(1)
+            pop = match.group(2) if match.group(2) else 'W'
+            if homeworld + '?' in initial_codes:
+                homeworld += "?"
+                pop = '0'
+            elif homeworld + 'X' in initial_codes:
+                pop = 'X'
+                homeworld += str(pop)
+            elif homeworld + str(pop) in initial_codes:
+                homeworld += str(pop)
+        sophont = self._process_sophont_homeworld(code, pop, direct_homeworld=True)
+        self.codes = [code for code in self.codes if code != homeworld]
+        if sophont not in self.codes:
+            self.codes.append(sophont)
+
+    def _process_homeworld_new(self, homeworld, homeworlds_found):
```

Comment:
The codes for `[<name>]` are for the major race home worlds. If this is for just those, perhaps a better name.

Reply:
so it's ```[major race]``` and ```(minor race)``` then?

Zhdant (Zhdant 2719) , judging from the below, doesn't seem to have _any_ sophont codes:
```2719 Zhdant               A6547C8-F Ag An Cx                            - KM - 811   Zh K0 V           ```

AIUI, it should add ```[Zhodani]``` - how would the Qiknavrat (ie, Chirper) dieoff be listed?  ```Di(Chirper)``` ?